### PR TITLE
Generate max coding shreds when necessary

### LINKED ISSUE #7902: Recover the good parts of "Generate MAX_DATA_SHREDS_PER_FEC_BLOCK coding shreds for each FEC block"
#7474 was reverted for contributing to a bandwidth usage spike on the idling SLP 1.1 cluster.  Let's re-assess this fix and consider which parts of it we want to keep, if any, while avoiding the bandwidth usage concern.

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
PR, that generated maximum coding shreds for all FEC blocks, was reverted. The FEC blocks with empty ticks also contained maximum number of coding shreds. This caused increase in network bandwidth usage for idling cluster.

However, there are instances where we want to generate more coding shreds than data shreds. The relevant portion of reverted PR needs to be resurrected, with more checks to limit when maximum coding shreds are generated.

#### Summary of Changes
Limited the conditions for generating MAX_DATA_SHREDS_PER_FEC_BLOCK coding shreds for FEC blocks with less number of data shreds. For example,

- The shreds generated from a `Vec<Entry>` got split into multiple FEC blocks. The current FEC block contains a part (most likely, tail) of the shred set.

Fixes #7902

### CODE REVIEW

File: ledger/src/shred.rs [hunk 1]
```diff
@@ -607,12 +609,25 @@ impl Shredder {
...
             // always generate at least 1 coding shred even if the fec_rate doesn't allow it
-            let num_coding = Self::calculate_num_coding_shreds(num_data as f32, fec_rate);
+            let shred_count = if max_coding_shreds {
+                MAX_DATA_SHREDS_PER_FEC_BLOCK as usize
+            } else {
+                num_data
+            };
+            let num_coding = Self::calculate_num_coding_shreds(shred_count as f32, fec_rate);
+            if num_coding > num_data {
+                trace!(
```
Comment (carllin) [inline ledger/src/shred.rs]: This can be an assert because it should never happen right?

Comment (carllin) [inline ledger/src/shred.rs]: So the only time this trace should trigger is if we have some remaining shreds in a big batch? So for instance if I have 33 data shreds, then the first 32 go into the first batch, and then the last data shred will be padded with 32 coding shreds?

Can we add a test outlining this expected behavior?

Comment (pgarg66) [reply ledger/src/shred.rs]: @carllin , yes that's the case it's covering. Do you think it makes sense to generate max coding shreds even for batches that have < 32 data shreds. For example, say the TPS is such that every  batch has around 17 data shreds. Right now it'll generate that many coding shreds.

I'll add the test to cover the conditions..

Comment (pgarg66) [reply ledger/src/shred.rs]: Added the test

**Pull request merged**